src/lib.rs

```diff
@@ -21,56 +21,72 @@
-        unsafe {llvm_asm!("1: sbiw $0,1
-                      brne 1b"
-                     :
-                     : "w" (0)
-                     :
-                     :)}
-    }
-    unsafe {llvm_asm!("1: sbiw $0,1
-                      brne 1b"
-                 :
-                 : "w" (last_count)
-                 :
-                 :)}
-}
-
-///delay for N miliseconds
-/// # Arguments
-/// * 'ms' - an u32, number of milliseconds to busy-wait
+/// * 'ms' - a u32, number of milliseconds to busy-wait
+#[inline(always)]
 pub fn delay_ms(ms: u32) {
-    // microseconds
-    let us = ms * 1000;
-    delay_us(us);
+    let ticks: u64 = (u64::from(avr_config::CPU_FREQUENCY_HZ) * u64::from(ms)) / 4_000;
+    delay_impl(ticks);
 }
 
 ///delay for N microseconds
 /// # Arguments
-/// * 'ms' - an u32, number of microseconds to busy-wait
+/// * 'ms' - a u32, number of microseconds to busy-wait
+#[inline(always)]
 pub fn delay_us(us: u32) {
-    // picoseconds
-    let ps = us * 1000;
-    let ps_lp = 1000000000 / (avr_config::CPU_FREQUENCY_HZ / 4);
-    let loops = (ps / ps_lp) as u32;
-    delay(loops);
+    let ticks: u64 = (u64::from(avr_config::CPU_FREQUENCY_HZ) * u64::from(us)) / 4_000_000;
+    delay_impl(ticks);
 }
 
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+///delay for N*4 clock cycles, not including setup overhead
+/// # Arguments
+/// * 'count' - a u32, number of iterations to busy-wait (4 cycles per iteration)
+#[inline(always)]
+pub fn delay(count: u32) {
+    delay_impl(u64::from(count));
+}
+
+
+/// Internal function to implement a variable busy-wait loop. 
+/// Delays for up up to 2^48 iterations, not including setup overhead.
+/// # Arguments
+/// * 'count' - a u64, the number of times to iterate the loop.
+/// The top 16 bits of count are ignored.
+#[inline(always)]
+fn delay_impl(count: u64) {
+    // Casting truncates, so this is equivalent to `(count & 0xFFFF) as u16`
+    let remaining_count: u16 = count as u16;
+    // if remaining_count is zero, calling delay_loop_4_cycles would delay for 2^16 cycles,
+    // so we don't loop in this case
+    if remaining_count != 0 {
+        delay_loop_4_cycles(remaining_count);
+    }
+    // Casting truncates, so this is equivalent to `((count >> 16) & 0xFFFF_FFFF) as u32`
+    let outer_count: u32 = (count >> 16) as u32;
+    for _ in 0..outer_count {
+        // Each loop through should be 4 cycles, so we're getting 262,144 cycles per outer loop.
+        delay_loop_4_cycles(0);
     }
 }
 
+/// Internal function to implement a 16-bit busy-wait loop in assembly.
+/// Delays for 4 cycles per iteration, not including setup overhead.
+/// Up to 2^16 iterations (the value 2^16 would have to be passed as 0).
+/// # Arguments
+/// * 'count' - a u16, the number of times to cycle the loop.
+#[inline(always)]
+#[allow(unused_variables, unused_mut, unused_assignments)]
+fn delay_loop_4_cycles(mut count: u16) {
```

Comment:
I had to adapt my own code to the new `asm` macro recently.  I found all the magical register naming in the Rust compiler source code at https://github.com/rust-lang/rust/blob/9ad5d82f822b3cb67637f11be2e65c5662b66ec0/compiler/rustc_target/src/asm/avr.rs.

The documentation of the `asm!` macro itself can be found there: https://doc.rust-lang.org/nightly/reference/inline-assembly.html.

Volatile is now the default. So I believe the following code should be the correct translation.
Even though we do not care of the final value of `_count`, we still must declare the register pair as both input and output. The hardware is mutating the register pair, and the compiler should know about this.

```rust
  /// 4 cycles per iteration.
  #[inline(always)]
  pub fn busy_loop_4_cycles(count: u16) {
      let mut _count = count; // Quiet the linter.
      unsafe {
          asm!(
              "1: sbiw {regpair}, 1",
              "brne 1b",
              regpair = inout(reg_iw) _count,
          )
      }
  }
```

Reply:
@bombela please make a merge request for it.